src/IO/PigzDeflatingWriteBuffer.cpp

```diff
@@ -0,0 +1,221 @@
+#include <list>
+
+#include <IO/PigzDeflatingWriteBuffer.h>
+#include <Common/Exception.h>
+#include <Common/ThreadPool.h>
+
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int ZLIB_DEFLATE_FAILED;
+}
+
+PigzDeflatingWriteBuffer::PigzDeflatingWriteBuffer(
+    std::unique_ptr<WriteBuffer> out_,
+    int compression_level_,
+    std::string filename_,
+    size_t buf_size,
+    char * existing_memory,
+    size_t alignment)
+    : WriteBufferWithOwnMemoryDecorator(std::move(out_), buf_size, existing_memory, alignment)
+    , compression_level(compression_level_)
+    , filename(filename_)
+{
+}
+
+const size_t MAXP2 = UINT_MAX - (UINT_MAX >> 1);
+const size_t BLOCK = 131072;
+
+void PigzDeflatingWriteBuffer::nextImpl()
+{
+    if (!offset())
+        return;
+
+    writeHeader();
+
+    auto *in_buf = reinterpret_cast<unsigned char *>(working_buffer.begin());
+    size_t in_len = offset();
+    uint64_t check = crc32_z(0L, Z_NULL, 0);
+
+    size_t cnt_blocks = in_len / BLOCK + bool(in_len % BLOCK);
+    std::vector<CompressedBuf> results(cnt_blocks);
+    std::vector<size_t> checks(cnt_blocks);
+    std::vector<size_t> blocks(cnt_blocks);
+
+    ThreadPool pool(15);
```

Comment:
Running thread pool of fixed inside write buffer seems a bit wrong. The number of threads should be managed outside the write buffer because we can't control the number of threads for query globally in that case. I don't have a ready solution, but we need to figure out how it's handled in other existing methods.

Reply:
> we need to figure out how it's handled in other existing methods

UPD: all current method do not use multiple threads.

One option is to use `IOThreadPool` from https://github.com/ClickHouse/ClickHouse/pull/35150 .